src/Crypto/WebAuthn/Operations/Attestation/TPM.hs

```diff
@@ -0,0 +1,589 @@
+
+  asfVerify
+    _
+    Statement {..}
+    M.AuthenticatorData {adRawData = M.WithRaw adRawData, ..}
+    clientDataHash = do
+      -- 1. Verify that attStmt is valid CBOR conforming to the syntax defined
+      -- above and perform CBOR decoding on it to extract the contained fields.
+      -- NOTE: This is done during decoding
+
+      -- 2. Verify that the public key specified by the parameters and unique
+      -- fields of pubArea is identical to the credentialPublicKey in the
+      -- attestedCredentialData in authenticatorData.
+      let pubKey = M.acdCredentialPublicKey adAttestedCredentialData
+      extractedPubKey <- maybe (Left VerificationErrorExtractingPublicKey) pure $ extractPublicKey pubArea
+      unless (pubKey == extractedPubKey) $ Left VerificationErrorCredentialKeyMismatch
+
+      -- 3. Concatenate authenticatorData and clientDataHash to form attToBeSigned.
+      let attToBeSigned = adRawData <> BA.convert (M.unClientDataHash clientDataHash)
+
+      -- 4. Validate that certInfo is valid:
+      -- 4.1 Verify that magic is set to TPM_GENERATED_VALUE.
+      let magic = tpmsaMagic certInfo
+      unless (magic == tpmGeneratedValue) . Left $ VerificationErrorInvalidMagicNumber magic
+
+      -- 4.2 Verify that type is set to TPM_ST_ATTEST_CERTIFY.
+      let typ = tpmsaType certInfo
+      unless (typ == tpmStAttestCertify) . Left $ VerificationErrorInvalidType typ
+
+      -- 4.3 Verify that extraData is set to the hash of attToBeSigned using
+      -- the hash algorithm employed in "alg".
+      attHash <- case alg of
+        PublicKey.COSEAlgorithmIdentifierRS1 -> pure $ BA.convert (hash attToBeSigned :: Digest SHA1)
+        PublicKey.COSEAlgorithmIdentifierRS256 -> pure $ BA.convert (hash attToBeSigned :: Digest SHA256)
+        PublicKey.COSEAlgorithmIdentifierRS384 -> pure $ BA.convert (hash attToBeSigned :: Digest SHA384)
+        PublicKey.COSEAlgorithmIdentifierRS512 -> pure $ BA.convert (hash attToBeSigned :: Digest SHA512)
+        PublicKey.COSEAlgorithmIdentifierES256 -> pure $ BA.convert (hash attToBeSigned :: Digest SHA256)
+        PublicKey.COSEAlgorithmIdentifierES384 -> pure $ BA.convert (hash attToBeSigned :: Digest SHA384)
+        PublicKey.COSEAlgorithmIdentifierES512 -> pure $ BA.convert (hash attToBeSigned :: Digest SHA512)
+        _ -> Left VerificationErrorUnknownHashFunction
+      let extraData = tpmsaExtraData certInfo
+      unless (attHash == extraData) . Left $ VerificationErrorHashMismatch attHash extraData
+
+      -- 4.5 Verify that attested contains a TPMS_CERTIFY_INFO structure as
+      -- specified in [TPMv2-Part2] section 10.12.3, whose name field contains
+      -- a valid Name for pubArea, as computed using the algorithm in the
+      -- nameAlg field of pubArea using the procedure specified in
+      -- [TPMv2-Part1] section 16.
+      pubAreaHash <- case tpmtpNameAlg pubArea of
+        TPMAlgSHA1 -> pure (BA.convert (hash pubAreaRaw :: Digest SHA256) :: BS.ByteString)
```

Comment:
This should arguably be `Digest SHA1` instead, do we not have a test for this case?